crypto/mlkem/mlkem-c.inc

```diff
@@ -0,0 +1,1212 @@
+    /*
+     * On Aarch64, omitting a |value_barrier_u16| results in a 2x speedup of
+     * ML-KEM overall and Clang still produces constant-time code using `csel`. On
+     * other platforms & compilers on godbolt that we care about, this code also
+     * produces constant-time output.
+     */
+    return (mask & x) | (~mask & subtracted);
+}
+
+/*
+ * constant time reduce x mod kPrime using Barrett reduction. x must be less
+ * than kPrime + 2 * kPrime^2.  This is sufficient to reduce a product of
+ * two already reduced u_int16 values, in fact it is sufficient for each
+ * to be less than 2^12, because (kPrime * (2 * kPrime + 1)) > 2^24.
+ */
+static uint16_t reduce(uint32_t x)
+{
+    uint64_t product = (uint64_t)x * kBarrettMultiplier;
+    uint32_t quotient = (uint32_t)(product >> kBarrettShift);
+    uint32_t remainder = x - quotient * kPrime;
+
+    assert(x < kPrime + 2u * kPrime * kPrime);
+    return reduce_once(remainder);
+}
+
+/*
+ * FIPS 203, Section 4.3, Algoritm 9: NTT.  In-place number theoretic transform
+ * of a given scalar.  Note that MLKEM's kPrime 3329 does not have a 512th root
+ * of unity, so this transform leaves off the last iteration of the usual FFT
+ * code, with the 128 relevant roots of unity being stored in |kNTTRoots|. This
+ * means the output should be seen as 128 elements in GF(3329^2), with the
+ * coefficients of the elements being consecutive entries in |s->c|.
+ */
+static void scalar_ntt(scalar *s)
+{
+    int offset = DEGREE;
+    int k, step, i, j;
+    uint32_t step_root;
+    uint16_t odd, even;
+
+    /*
+     * `int` is used here because using `size_t` throughout caused a ~5% slowdown
+     * with Clang 14 on Aarch64.
+     */
+    for (step = 1; step < DEGREE / 2; step <<= 1) {
+        offset >>= 1;
+        k = 0;
+        for (i = 0; i < step; i++) {
+            step_root = kNTTRoots[i + step];
+            for (j = k; j < k + offset; j++) {
+                odd = reduce(step_root * s->c[j + offset]);
+                even = s->c[j];
+                s->c[j] = reduce_once(odd + even);
+                s->c[j + offset] = reduce_once(even - odd + kPrime);
+            }
+            k += 2 * offset;
+        }
+    }
+}
+
+/* In-place NTT transform of a vector */
+static void vector_ntt(vector *a)
+{
+    int i;
+
+    for (i = 0; i < RANK; i++)
+        scalar_ntt(&a->v[i]);
+}
+
+/*
+ * FIPS 203, Section 4.3, Algoritm 10: NTT^(-1).  In-place inverse number
+ * theoretic transform of a given scalar, with pairs of entries of s->v being
+ * interpreted as elements of GF(3329^2). Just as with the number theoretic
+ * transform, this leaves off the first step of the normal iFFT to account for
+ * the fact that 3329 does not have a 512th root of unity, using the
+ * precomputed 128 roots of unity stored in |kInverseNTTRoots|.
+ *
+ * FIPS 203, Algorithm 10, performs this transformation in a slightly different
+ * manner, using the same kNTTRoots table as the forward NTT transform.
+ */
```

Comment:
Would there be any benefit in making this match the FIPS 203 algorithm 10 ? That'd at least avoid the need for a second table, have not looked at whether this is more run-time efficient as written?